cmd/gpu_plugin/rm/gpu_plugin_resource_manager.go

```diff
@@ -0,0 +1,414 @@
+					name:                podRes.Name,
+					allocatedContainers: numContainersAllocated,
+					allocationTargetNum: allocationTargetNum,
+				}
+				candidates = append(candidates, candidate)
+			}
+		}
+	}
+
+	return candidates, nil
+}
+
+func (rm *ResourceManager) SetDevTree(devTree dpapi.DeviceTree) {
+	rm.mutex.Lock()
+	defer rm.mutex.Unlock()
+	rm.deviceInfos = map[string]DeviceInfoSubset{}
+
+	if deviceIDMap, ok := devTree[deviceType]; ok {
+		for deviceID, info := range deviceIDMap {
+			if strings.HasSuffix(deviceID, "-0") {
+				rm.deviceInfos[deviceID] = DeviceInfoSubset{
+					nodes:  info.Nodes(),
+					mounts: info.Mounts(),
+					envs:   info.Envs(),
+				}
+			}
+		}
+	}
+}
+
+func (rm *ResourceManager) createAllocateResponse(cards []string) (*pluginapi.AllocateResponse, error) {
+	rm.mutex.Lock()
+	defer rm.mutex.Unlock()
+
+	allocateResponse := pluginapi.AllocateResponse{}
+	cresp := pluginapi.ContainerAllocateResponse{}
+
+	for _, card := range cards {
+		newDeviceID := card + "-0"
+
+		dev, ok := rm.deviceInfos[newDeviceID]
+		if !ok {
+			klog.Warningf("No device info for %v, using default allocation method devices", newDeviceID)
+			return nil, dpapi.UseDefaultAllocateMethodError{}
+		}
+
+		// add new devices
+		nodes := dev.nodes
+		for i := range dev.nodes {
+			cresp.Devices = append(cresp.Devices, &nodes[i])
+		}
+
+		// add new mounts
+		mounts := dev.mounts
+		for i := range mounts {
+			cresp.Mounts = append(cresp.Mounts, &mounts[i])
+		}
+
+		for key, value := range dev.envs {
+			if cresp.Envs == nil {
+				cresp.Envs = make(map[string]string)
+			}
+			cresp.Envs[key] = value
+		}
+	}
+
+	allocateResponse.ContainerResponses = append(allocateResponse.ContainerResponses, &cresp)
+
+	return &allocateResponse, nil
+}
+
+func numGPUUsingContainers(pod *v1.Pod) int {
+	num := 0
+	for _, container := range pod.Spec.Containers {
+		for reqName, quantity := range container.Resources.Requests {
+			resourceName := reqName.String()
+			if resourceName == fullResourceName {
+				value, _ := quantity.AsInt64()
+				if value > 0 {
+					num++
+					break
+				}
+			}
+		}
+	}
+	return num
+}
+
+// containerCards returns the cards to use for a single container.
+// gpuUsingContainerIndex 0 == first gpu-using container in the pod.
+func containerCards(pod *v1.Pod, gpuUsingContainerIndex int) []string {
+	fullAnnotation := pod.Annotations[gasCardAnnotation]
+	cardLists := strings.Split(fullAnnotation, "|")
+	klog.V(3).Infof("%s:%v", fullAnnotation, cardLists)
+
+	i := 0
+	for _, cardList := range cardLists {
+		cards := strings.Split(cardList, ",")
+		if len(cards) > 0 && len(cardList) > 0 {
+			if gpuUsingContainerIndex == i {
+				klog.V(3).Infof("Cards for container nr %v are %v", gpuUsingContainerIndex, cards)
+				return cards
+			}
+			i++
+		}
+	}
+	klog.Warningf("couldn't find cards for gpu using container index %v", gpuUsingContainerIndex)
+	return nil
+}
+
+func getClientset() *kubernetes.Clientset {
```

Comment:
This function would be shorter if it returned `error` without logging it. Then the code handling the error could log all possible errors in one place.
```suggestion
func getClientset() (*kubernetes.Clientset, error) {
```

Reply:
done